File: java/src/main/java/io/cucumber/prettyformatter/AttachmentFormatter.java
package io.cucumber.prettyformatter;

import io.cucumber.messages.types.Attachment;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;

import static io.cucumber.prettyformatter.Theme.Element.ATTACHMENT;

final class AttachmentFormatter {

    private final int indentation;

    private AttachmentFormatter(int indentation) {
        this.indentation = indentation;
    }

    static Builder builder() {
        return new Builder();
    }

    void formatTo(Attachment attachment, LineBuilder lineBuilder) {
        switch (attachment.getContentEncoding()) {
            case BASE64 -> formatBase64Attachment(attachment, lineBuilder);
            case IDENTITY -> formatTextAttachment(attachment, lineBuilder);
        }
    }

    private void formatBase64Attachment(Attachment attachment, LineBuilder lineBuilder) {
        int bytes = (attachment.getBody().length() / 4) * 3;
        String line = attachment.getFileName().isPresent() //
                ? "Embedding %s [%s %d bytes]".formatted(attachment.getFileName().get(), attachment.getMediaType(), bytes) //
                : "Embedding [%s %d bytes]".formatted(attachment.getMediaType(), bytes);

        lineBuilder.indent(indentation)
                .append(ATTACHMENT, line)
                .newLine();
    }

    private void formatTextAttachment(Attachment attachment, LineBuilder lineBuilder) {
        // Prevent interleaving when multiple threads write to System.out
        try (BufferedReader lines = new BufferedReader(new StringReader(attachment.getBody()))) {
            String line;
            while ((line = lines.readLine()) != null) {
                lineBuilder.indent(indentation)
                        .append(ATTACHMENT, line)
                        .newLine();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    static final class Builder {
        private int indentation = 0;

        private Builder() {
        }

        Builder indentation(int indentation) {
            this.indentation = indentation;
            return this;
        }

        AttachmentFormatter build() {
            return new AttachmentFormatter(indentation);
        }
    }
}

File: cpp/src/cucumber/pretty-formatter/AttachmentFormatter.cpp
#include "cucumber/pretty-formatter/AttachmentFormatter.hpp"
#include "cucumber/messages/Attachment.hpp"
#include "cucumber/messages/AttachmentContentEncoding.hpp"
#include "cucumber/pretty-formatter/LineBuilder.hpp"
#include "cucumber/pretty-formatter/Theme.hpp"
#include <cstddef>
#include <fmt/format.h>
#include <memory>
#include <sstream>
#include <string>

namespace cucumber::pretty_formatter
{
    AttachmentFormatter::AttachmentFormatter(std::size_t indent)
        : indent{ indent }
    {}

    void AttachmentFormatter::Format(LineBuilder& lineBuilder, const std::shared_ptr<const messages::Attachment>& attachment)
    {
        if (attachment->contentEncoding == messages::AttachmentContentEncoding::BASE64)
        {
            FormatBase64(lineBuilder, attachment);
        }
        else
        {
            FormatText(lineBuilder, attachment);
        }
    }

    void AttachmentFormatter::FormatBase64(LineBuilder& lineBuilder, const std::shared_ptr<const messages::Attachment>& attachment) const
    {
        const auto bytes = (attachment->body.size() / 4) * 3;
        const auto filename = attachment->fileName.has_value() ? attachment->fileName.value() + " " : "";

        lineBuilder.Indent(indent)
            .Append(Theme::Element::attachment, fmt::format("Embedding {}[{} {} bytes]", filename, attachment->mediaType, bytes))
            .NewLine();
    }

    void AttachmentFormatter::FormatText(LineBuilder& lineBuilder, const std::shared_ptr<const messages::Attachment>& attachment) const
    {
        std::istringstream stream{ attachment->body };
        for (std::string line; std::getline(stream, line);)
        {
            lineBuilder.Indent(indent).Append(Theme::Element::attachment, line).NewLine();
        }
    }
}
